### PR TITLE
CALS-5542 Physical address toggle bug and refactor CALS-5542

### PR DESCRIPTION
**Pull request by lauherkcambria**
#### Which User story does this relate to?
- https://osi-cwds.atlassian.net/browse/CALS-5542

#### Brief feature description
- fix bug related to physical address  toggle for mailing address in residence card. refactor address card/ address component and their references.

#### Does this PR block/effect other stories?
-effected story: https://osi-cwds.atlassian.net/browse/CALS-5018
#### Do any commands needs to be run in order to run the project with this PR?
-no

#### Code coverage percent diff as a result of this PR?
82.05% Statements
68.14% Branches 
78.01% Functions
82.7% Lines

### CODE REVIEW

**Comment by Nasir2712** [comment]
I have two questions-
1) Are we still not breaking ```addressFields``` object into seperate zip, state and street address?
2) Do we still need ```addressCard.js``` since it the entire addressCard is just ```addressComponent.js```

File: app/javascript/components/rfa_forms/addressComponent.js [hunk 1]
```diff
@@ -1,6 +1,8 @@
 import React from 'react'
 import PropTypes from 'prop-types'
 import {CommonAddressComponent} from 'react-wood-duck'
+import {physicalAddressType, mailingAddressType} from 'constants/constants'
```
[the remark this answers is not in the document]
Comment (lauherkcambria) [reply app/javascript/components/rfa_forms/addressComponent.js]: that's a good idea. 👍

File: app/javascript/constants/constants.js [hunk 2]
```diff
@@ -94,4 +94,7 @@ const FacilityTypes = Object.freeze({
   '733': 'SHORT TERM RESIDENTIAL THERAPEUTIC PROGRAM - GH'
 })
 
-export {CountyList, yesNo, FacilityTypes}
+const physicalAddressType = 'Residential'
+const mailingAddressType = 'Mailing'
+
+export {CountyList, yesNo, FacilityTypes, physicalAddressType, mailingAddressType}
```
[the remark this answers is not in the document]
Comment (lauherkcambria) [reply app/javascript/constants/constants.js]: ok

File: app/javascript/rfa_forms/rfa01a_edit_view/addressCard.js [hunk 3]
```diff
@@ -1,114 +1,68 @@
...
-
 export default class AddressCard extends React.Component {
   constructor (props) {
     super(props)
-    this.state = {
-      suggestions: []
-    }
-    this.checkAddressType = this.checkAddressType.bind(this)
+    this.onMailingAddressChange = this.onMailingAddressChange.bind(this)
+    this.onPhysicalAddressChange = this.onPhysicalAddressChange.bind(this)
```
[the remark this answers is not in the document]
Comment (lauherkcambria) [reply app/javascript/rfa_forms/rfa01a_edit_view/addressCard.js]: thanks! 😄

File: app/javascript/rfa_forms/rfa01a_edit_view/addressCard.js [hunk 4]
```diff
@@ -1,114 +1,68 @@
...
-            </div>
-            <div className={hiddenMailingSameAsPhysical}>
-              <AddressComponent
+          <AddressComponent
+            stateTypes={this.props.stateTypes}
+            addressTitle='Physical Address'
+            id='street_address'
+            addressType={physicalAddressType}
+            addressFields={this.props.physicalAddress}
+            mailingAddress={this.props.mailingAddress}
```
[the remark this answers is not in the document]
Comment (lauherkcambria) [reply app/javascript/rfa_forms/rfa01a_edit_view/addressCard.js]: good question. the api response is expecting residence  addresses in the form of 

` "residence": {
    "addresses": [
      {
        "street_address": "4210 Ardwell Way",
        "zip": "95823",
        "city": "Sacramento",
        "state": {
          "value": "California",
          "id": "CA"
        },
        "type": {
          "value": "Residential",
          "id": 1
        }
      },
      {
        "street_address": "4230 Ardwell Way",
        "zip": "95823",
        "city": "Sacramento",
        "state": {
          "value": "California",
          "id": "CA"
        },
        "type": {
          "value": "Mailing",
          "id": 3
        }
      }
    ]`

so if we are to move the onSelection() method out of the address card and into  the common address component we have to have logic for setting the blank values for the opposite card we are on.

File: app/javascript/components/rfa_forms/addressComponent.js [hunk 5]
```diff
@@ -89,7 +109,7 @@ AddressComponent.propTypes = {
 }
 
 AddressComponent.defaultProps = {
-  addressType: '',
+  addressType: physicalAddressType,
```
[the remark this answers is not in the document]
Comment (lauherkcambria) [reply app/javascript/components/rfa_forms/addressComponent.js]: good catch

**Comment by lauherkcambria** [comment]
@Nasir2712 

thank you for your comments!

1.  on refactor, it didn't make much sense to do this given we would have to transform it back into an object anyway in order to use commonAddressComponent into react-wood-duck. 
2. in a way we are doing what you are suggesting - we still need addressCard.js as it is handling the functions for on physical address change and mailing address change in the rfa01a form, but address card is only on this form, and address component is used elsewhere.

File: app/javascript/components/rfa_forms/addressComponent.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (lauherkcambria) [reply app/javascript/components/rfa_forms/addressComponent.js]: we don't need a default prop of addresses here tho, cause we take care of the defaults before it even gets here.

File: app/javascript/rfa_forms/rfa01a_edit_view/addressCard.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (ratneshraval) [reply app/javascript/rfa_forms/rfa01a_edit_view/addressCard.js]: Thanks for clarification. I realized after I read the code in detail.

**Comment by ratneshraval** [comment]
@lauherkcambria can you put story number in PR title?

File: app/javascript/components/rfa_forms/addressComponent.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (ratneshraval) [reply app/javascript/components/rfa_forms/addressComponent.js]: I didn't see prototype declared.

File: app/javascript/rfa_forms/rfa01a_edit_view/residenceCardsMain.jsx [hunk 6]
```diff
@@ -3,44 +3,37 @@ import Immutable from 'immutable'
...
-      newData = newData.set(key, value)
-      newData = newData.set(hiddenKey, hiddenDefaultValue)
-      this.props.setParentState('residence', newData.toJS())
+    if (value === 'true') {
+      let residence = Immutable.fromJS(this.props.residence)
+      residence = residence.set(key, value)
+      if (hiddenKey === 'addresses') {
+        // if value from the conditional is true, we want to clear the entered data
+        // in Mailing address. need to set the key/value from  regardless.
+        let index = this.props.residence.addresses.findIndex(o => o.type.value === 'Mailing')
```
Comment (ratneshraval) [inline app/javascript/rfa_forms/rfa01a_edit_view/residenceCardsMain.jsx]: `const` here?

File: app/javascript/components/rfa_forms/addressComponent.js [hunk 7]
```diff
@@ -12,30 +14,47 @@ export default class AddressComponent extends React.Component {
...
+    const blankAddressFields = (addressType === physicalAddressType)
+      ? blankPhysicalAddress : blankMailingAddress
 
-    let addressObj = Immutable.fromJS(this.props.addressFields)
-    addressObj = addressObj.update(x => updateSuggetions)
-    this.props.onSelection(addressObj)
+    autoFillData.type = blankAddressFields.type
+
+    let addresses = addressType === physicalAddressType
+      ? Immutable.List([autoFillData, this.props.mailingAddress].filter(n => n))
```
Comment (ratneshraval) [inline app/javascript/components/rfa_forms/addressComponent.js]: What does the `.filter(n => n)` do?

Comment (lauherkcambria) [reply app/javascript/components/rfa_forms/addressComponent.js]: removes null, undefined and "" from the array

**Pull request merged**